### DISCUSSION TITLE
Clarification Request regarding the Executor component...

### DISCUSSION THREAD

**Discussion by michaelr-tradesy** [badoo/MVIKotlin#244] [Q&A] [answered]
Hi...

I'm further exploring the MVIKotlin framework, a continuation of [this previous discussion](https://github.com/arkivanov/MVIKotlin/discussions/237). This time, I'm attempting to eliminate the need to provide a reducer to my custom store. I'm continuing that Calculator example provided in your [tutorial / guide](https://arkivanov.github.io/MVIKotlin/store.html). Below is the source code as it currently stands:

### Calculator Controller
```
interface CalculatorController {
    val flow: MutableStateFlow<Long>
    fun onIncrement()
    fun onDecrement()
}

@ExperimentalStateKeeperApi
@ExperimentalInstanceKeeperApi
class DefaultCalculatorController(
    lifecycle: Lifecycle,
    instanceKeeper: InstanceKeeper,
    private val stateKeeper: StateKeeper<CalculatorStore.State>?
) :
    CalculatorController, BaseMviView<CalculatorView.Model, CalculatorView.Event>() {

    private val store: CalculatorStore
    private val storeFactoryInstance: StoreFactory
    private var binder: Binder? = null

    override val flow = MutableStateFlow(0L)

    init {
        lifecycle.doOnStartStop(::onStart, ::onStop)

        storeFactoryInstance =
            LoggingStoreFactory(delegate = TimeTravelStoreFactory(fallback = DefaultStoreFactory))
        val calculatorStore = {
            CalculatorStoreFactory(storeFactoryInstance).create(stateKeeper)
        }
        store = instanceKeeper.getStore(calculatorStore)

        binder =
            com.arkivanov.mvikotlin.extensions.reaktive.bind {
                store.states.bindTo {
                    flow.value = it.value
                    println("store.states(): New State=($it)...")
                }
                store.labels.bindTo {
                    flow.value = it.value
                    println("store.labels(): New Label=($it)...")
                }
            }
    }

    private fun onStart() {
        binder?.start()
    }

    private fun onStop() {
        binder?.stop()
    }

    override fun onIncrement() {
        store.accept(CalculatorStore.Intent.Increment)
    }

    override fun onDecrement() {
        store.accept(CalculatorStore.Intent.Decrement)
    }
}
```

### Calculator Store Factory
```
@ExperimentalStateKeeperApi
internal class CalculatorStoreFactory(private val storeFactory: StoreFactory) {

    private val initialValue = 0L

    fun create(stateKeeper: StateKeeper<CalculatorStore.State>?): CalculatorStore =
        // Make sure the values of the generics are consistent with the store.
        object : CalculatorStore, Store<CalculatorStore.Intent, CalculatorStore.State, CalculatorStore.State> by storeFactory.create(
            name = "CounterStore",
            initialState = CalculatorStore.State(initialValue),
            executorFactory = executor,
            reducer = { it }
        ){
        }.also {
            stateKeeper?.register {
                // We can reset any transient state here
                it.state.copy(value = initialValue)
            }
        }

    private val executorImpl = ExecutorImpl()
    private val executor: () -> Executor<CalculatorStore.Intent, Nothing, CalculatorStore.State, CalculatorStore.State, CalculatorStore.State> = {
        executorImpl
    }

    private class ExecutorImpl
        : SuspendExecutor<CalculatorStore.Intent, Nothing, CalculatorStore.State, CalculatorStore.State, CalculatorStore.State>() {
        override suspend fun executeIntent(intent: CalculatorStore.Intent, getState: () -> CalculatorStore.State) =
            when (intent) {
                is CalculatorStore.Intent.Increment -> {
                    val output = CalculatorStore.State(getState().value + 1)
                    broadcast(output)
                }
                is CalculatorStore.Intent.Decrement -> {
                    val output = CalculatorStore.State(getState().value - 1)
                    broadcast(output)
                }
            }

        private fun broadcast(output: CalculatorStore.State) {
            //Dispatch the `Result` to the [Reducer], which will `update the `State`
            dispatch(output)
            //Immediately send the `Label` to the [Store] for publication
            publish(output)
        }
    }
}
```

fyi: Applies to the **CalculatorController**: 
* I see the framework provides extension methods, so I opted to use those to bind back to my controller. 
* I also see that I can publish to the story by way of ```store::accept()```, so I opted to do that as well.

I notice that if I want changes to my state to be applied internally, I must provide a reducer in my custom store. Otherwise, the state of my store will remain unchanged.

I'm currently reviewing the TimeTravelStoreImpl.kt, the **EventProcessor** inner class, the **process()** method. My goal is to change the state of my store, and publish a label, without the need for using a Reducer (aka, process a result). 

I also took a look at the StoreFactory.kt, also in the library. It looks like we have a bypassReducer defined in the event I do not provide my own reducer. It looks like that reducer passes the initial state that I provided in my custom store. 

Here is the question that I have: Shouldn't the bypass reducer pass the state that has been provided as the parameter? In other words, shouldn't this:
```
 private companion object {
        private val bypassReducer: Reducer<Any, Any> =
            object : Reducer<Any, Any> {
                override fun Any.reduce(result: Any): Any = this
            }
    }
```
be this instead?
```
 private companion object {
        private val bypassReducer: Reducer<Any, Any> =
            object : Reducer<Any, Any> {
                override fun Any.reduce(result: Any): Any { return result ?: this }
            }
    }
```
???

Are you able to shed some light as to what I'm currently seeing?

One more thing: Did you have a specific intention in mind with regards to dispatching Results vs publishing Labels?

Thanks for the clarification in advance. I like what you've done here.

Edit: 
The anticipated behavior is that, given the initial value of 0, that I'm successfully able to increment and decrement as much as I want. Instead, if I do not provide a reducer, the actual behavior is that I am unable to change the state outside of the range -1 and 1.

**Comment by arkivanov** [comment, marked as answer]
Thanks for your question!

1. Regarding the `bypassReducer`. The `Reducer` interface has two generic type parameters: `Result` and `State`. The `Reducer` applies a `Result` to the current `State` and returns a new `State`. It can not return the `result` value, because it has different type.
2. By avoiding the `Reducer` and manipulating the `State` in the `Executor` you are abusing the library. I know, it might feel less boilerplate-ish, but it moves the responsibility of the `Reducer` to the `Executor`. The `Reducer` is designed to be a pure synchronous function, which responsibility is to only apply a `Result` to the current `State` and return a new `State`. The `Executor` can be asynchronous and can produce side effects. Such separation is also useful during debugging, especially time travelling.
3. `Labels` are side effects of a `Store`. They usually represent one time events. I think you don't need to publish `State` via `Label`. You can observe states as follows: `myStore.states.distinctUntilChanged()`.

**Reply by michaelr-tradesy** [reply]
Sorry for the long delay... understood...

I revisited the documentation. Because I followed the guide in its entirety, I overlooked [this detail](https://arkivanov.github.io/MVIKotlin/store.html):

```
The only component we need is the Reducer. It accepts Intents and modifies the State by incrementing or decrementing its value. The factory function create() uses the StoreFactory which is passed as a dependency.
```
Since the reducer can also accept an Intent, I decided to remove the executor from my code.

**Reply by arkivanov** [reply]
Correct. If you don't need any async actions, then you don't need the `Executor`.
